source/server/connection_handler_impl.cc

```diff
@@ -84,8 +84,19 @@ void ConnectionHandlerImpl::enableListeners() {
 
 void ConnectionHandlerImpl::ActiveTcpListener::removeConnection(ActiveTcpConnection& connection) {
   ENVOY_CONN_LOG(debug, "adding to cleanup list", *connection.connection_);
-  ActiveTcpConnectionPtr removed = connection.removeFromList(connections_);
+  ActiveConnections& active_connections = connection.active_connections_;
+  ActiveTcpConnectionPtr removed = connection.removeFromList(active_connections.connections_);
   parent_.dispatcher_.deferredDelete(std::move(removed));
+  // Only lookup if defer delete is needed
```

Comment:
Not sure how this connects with the line below. If there was one connection before in `connections_` prior to this call, it's now empty regardless of the deferred delete or not.

Reply:
Yes, the `connections_` which represents the connection list is empty. But it leave a garbage entry in the map `connections_by_tag_`. If we don't erase it, eventually `connections_by_tag_` will be full of entries with `{ (obsoletetag1, empty connections), (obsoletetag2, empty connections), (obsoletetag3, empty connections) }`

It is not a huge issue at this PR because this PR doesn't really support destruct a single filter chain. However, I decide to add this clean up in this PR because this piece of code should do no harm.

Reply:
I still don't get the connection with the comment "Only lookup if...". I don't see where the conditional is happening, maybe I'm slow :)

Reply:
Sorry for the misleading. Will update.
The `defer delete` is describing the defer delete 3 lines below. 
The comment will be `only look up the keyvalue when the below connections in the value is empty`